src/lazy/encoder/mod.rs

```diff
@@ -0,0 +1,565 @@
+
+/// A family of types that collectively comprise the writer API for an Ion serialization
+/// format. These types operate at the 'raw' level; they do not attempt to resolve symbols
+/// using the active symbol table.
+// Implementations of this trait are typically unit structs that are never instantiated.
+// However, many types are generic over some `E: LazyEncoder`, and having this trait
+// extend 'static, Sized, Debug, Clone and Copy means that those types can #[derive(...)]
+// those traits themselves without boilerplate `where` clauses.
+pub trait LazyEncoder<W: Write>: 'static + Sized + Debug + Clone + Copy {
+    // XXX: ^-- This is named 'Lazy' for symmetry with the `LazyDecoder`. In reality, there's nothing
+    //      lazy about it. We should revisit the Lazy* naming scheme, as eventually it will be the
+    //      only implementation of a reader and won't need the word 'Lazy' to distinguish itself.
+
+    /// A writer that serializes Rust values as Ion, emitting the resulting data to an implementation
+    /// of [`Write`].
+    type Writer: LazyRawWriter<W, Self>;
+
+    /// A single-use type that can emit an Ion value.
+    type ValueWriter<'a>: ValueWriter<'a, W, Self>
+    where
+        W: 'a;
+
+    /// A single-use type that can emit a sequence of annotations and then return a [`ValueWriter`].
+    type AnnotatedValueWriter<'a>: AnnotatedValueWriter<'a, W, Self>
+    where
+        W: 'a;
+
+    /// Allows the application to write a (potentially heterogeneously typed) list without necessarily
+    /// having all of its child values in memory.
+    type ListWriter<'a>: SequenceWriter<'a, W, Self>
+    where
+        W: 'a;
+
+    // TODO: Apply trait constraints to the following associated types
+
+    type SExpWriter<'a>;
+    type StructWriter<'a>;
+    type EExpressionWriter<'a>;
+}
+
+impl<W: Write> LazyEncoder<W> for TextEncoding_1_0 {
+    type Writer = LazyRawTextWriter_1_0<W>;
+    type ValueWriter<'a> = TextValueWriter_1_0<'a, W> where W: 'a;
+    type AnnotatedValueWriter<'a> = TextAnnotatedValueWriter_1_0<'a, W> where W: 'a;
+    type ListWriter<'a> = TextListWriter_1_0<'a, W> where W: 'a;
+    type SExpWriter<'a> = ();
+    type StructWriter<'a> = ();
+    type EExpressionWriter<'a> = ();
+}
+
+/// One-shot methods that take a (possibly empty) sequence of annotations to encode and return a ValueWriter.
+pub trait AnnotatedValueWriter<'a, W: Write, E: LazyEncoder<W>> {
+    /// Writes the provided annotations to the output stream and returns a [`ValueWriter`] that can
+    /// be used to serialize the value itself.
+    ///
+    /// If there are no annotations, use [`Self::no_annotations`] instead. This method will loop
+    /// over the empty sequence, incurring minor performance overhead while `no_annotations`
+    /// is a true no-op.
+    fn write_annotations<
+        SymbolType: AsRawSymbolTokenRef,
+        IterType: Iterator<Item = SymbolType> + Clone,
```

Comment:
🗺️ An earlier version of this used much more generics; I started spelling out what each generic was (e.g. `SymbolType` instead of `A`). I think I kind of like it?